### ISSUE TITLE
User Story: MVP Deploy

### ISSUE THREAD

**Issue by DeeMcCart** [DeeMcCart/CI_PP5_Jeweller#9]
EPIC:  #7 

As a **system owner** I can **deploy and re-deploy the system with MVP functionality** so that **the site operates as expected with MVP functionality**

### Assumptions or Pre-Requisites:
* Basic site setup is in place #1 
* Debug mode set to FALSE (note that this is now set using an environment variable so longer-term solution)
* First Deploy has been performed and any lessons learned incorporated
* Possibly some additional configuration settings needed in Heroku for MVP deploy (vs first deploy)

### Acceptance Criteria: (Must be completed before task is moved to 'Done')
- [x] Must be able to build app inHeroku
- [x] Must be able to run app once built

### Tasks - Build ElephantSQL Database
- [x] Start with ElephantSQL and verify database in place... YIKES not there **(first deployment took different approach)** 
- [x] Install psycopg2, dj_database_url - and freeze requirements 
- [x]  Export data from existing database:  python3 manage.py dumpdata products > products.json
- [x] Settings.py: import dj_database_url.  Commment out existing DATABASES = and replace with DATABASES= { default: dj_database_url(<my_database_url_from_elephant>)}
- [x] python3 manage.py showmigrations - should see a list of migrations to be applied to the database I am now connected to (the elephantsql database)
- [x] python3 manage.py migrate - to apply all the database structures to the new database
- [x] IF USING FIXTURES to load JSON data:  python3 manage.py loaddata categories;   python3 manage.py loaddata products; 
- [x] create database superuser(s)
- [x] revert database settings in settings.py, remove URL to new ElephantSQL database and save/commit

### Tasks - load Data into ElephantSQL database
- [x] Check migrations in ElephantSQL - open up the ElephantSQL console
- [x] Browser - table queries - auth user - execute - should see superuser(s) setup in previous task
- [ ] OOPs - just realised - becuase I'm not using JSON I will need to manually key the data into my Production database

### Tasks - Prepare to Deploy
- [x] Add Stripe fields to config variables in Heroku
- [x] If not already done, add a conditional statement into settings.py to link to appropriate database depending on env settings
- [x] Ensure gunicorn is installed to act as web server & freeze
- [x] set Heroku config DISABLE_COLLECTSTATIC to 1 to prevent Heroku from trying to collect static files when we deploy
- [x] create Procfile in root directory to tell Heroku to treat this as a python app/ create a web dyno:  **web: gunicorn jeweller.wsgi --log-file -**
- [x] settings.py ALLOWED_HOSTS = ... Heroku_deployment_name
- [x] commit changes and attempt to build from Heroku
- [ ] setup secret_key and retrieve it from the environment (refer to 'Deploying to Heroku')

### Tasks - Build & Test (without static) 
- [x] BUILD IN HEROKU:
- [x] Usual craic with deploy - build

### Tasks - Configure AWS for Static files
- [x] CONFIGURE AWS FOR STATIC FILES
- [x] Create bucket,set to public access
- [x] Properties tab - scroll to bottom - static - enable, set default index.html error.html, Save Changes (This gives a new endpoint accessible from internet)
- [x]  Permissions tab - scroll to CORS , set to below, then Save Changes
[{"AllowedHeaders": ["Authorization"],
"AllowedMethods": ["GET"],
"AllowedOrigins": ["*"],
"ExposeHeaders": []}]
- [x]  Permissions tab - Bucket Policy - Edit - Policy generator: type S3 Bucket policy; Allow all Principals *; Action: GetObject; ARN: (paste from other tab); AddStatement button; SaveChanges button; GeneratePolicy button; (copy to other tab); insert */ at end of Resource string; SaveChanges button.
- [x]  Permissions tab - ACL - Edit - Grant Public: ListAccess; override message; SaveChanges
### Tasks - SET AWS IAM (Identity and Access Management)
- [x] Create Group
- [x] Create Access Policy, attach policy to Group, 
- [x] Create user, attach user to Group
- [x] Generate keys
### Tasks - TRANSFER STATIC FILES FROM DEV TO PROD
- [x] Install boto and django-storages, freeze and add 'storages' to installed apps
- [x] Configure settings.py to use AWS for static and media storage
- [x] Create file custom_storages.py and add some content...
- [x] Commit, build deployment
- [ ] Test app

**Comment by DeeMcCart** [comment]
DMcC 09/02/24 Got instructions from tutor support how to export product dataset from test environment (create a json using dumpdata yaay)

**Comment by DeeMcCart** [comment]
DMcC 10/02/24 deploy-build done but lots of problems with static elements with leading '/' - about 6 of these uncovered and resolved.  Big problems with images not being retrieved.... possibly related to removing whitenoise just before deploy (possibly prevengint it from being searched as an image location?)
